File: src/main/java/com/camerapoints/ui/GroupPanel.java
package com.camerapoints.ui;

import com.camerapoints.CameraPointsPlugin;
import com.camerapoints.models.CameraPoint;
import com.camerapoints.models.CameraPointGroup;
import net.runelite.client.config.Keybind;
import net.runelite.client.ui.ColorScheme;
import net.runelite.client.ui.FontManager;
import net.runelite.client.ui.components.FlatTextField;
import net.runelite.client.util.ImageUtil;

import javax.swing.Box;
import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.border.CompoundBorder;
import javax.swing.border.EmptyBorder;
import javax.swing.border.LineBorder;
import javax.swing.border.MatteBorder;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Cursor;
import java.awt.Dimension;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class GroupPanel extends JPanel
{
	private final CameraPointsPlugin plugin;
	private final CameraPointGroup group;
	private final Runnable reloadFunction;

	private final ImageIcon addIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "add_icon.png"));
	private final ImageIcon addHoverIcon = new ImageIcon(ImageUtil.alphaOffset(addIcon.getImage(), -100));
	private final ImageIcon addPressedIcon = new ImageIcon(ImageUtil.alphaOffset(addIcon.getImage(), -50));
	private final ImageIcon deleteIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "delete_icon.png"));
	private final ImageIcon deleteHoverIcon = new ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin.class, "delete_icon_red.png"));
	private final ImageIcon deletePressedIcon = new ImageIcon(ImageUtil.alphaOffset(deleteHoverIcon.getImage(), -50));

	private final JCheckBox enabledBox = new JCheckBox();
	private final FlatTextField nameInput = new FlatTextField();
	private final JLabel saveLabel = new JLabel("Save");
	private final JLabel cancelLabel = new JLabel("Cancel");
	private final JLabel renameLabel = new JLabel("Rename");
	private final JLabel addPointLabel = new JLabel(addIcon);
	private final JButton prevKeybind = new JButton();
	private final JButton nextKeybind = new JButton();
	private final JLabel deleteLabel = new JLabel(deleteIcon);
	private final JPanel pointsContainer;

	public GroupPanel(CameraPointsPlugin plugin, CameraPointGroup group, Runnable reloadFunction)
	{
		this.plugin = plugin;
		this.group = group;
		this.reloadFunction = reloadFunction;

		setLayout(new BorderLayout());
		setBorder(new EmptyBorder(5, 5, 5, 5));
		setBackground(ColorScheme.DARKER_GRAY_COLOR);

		JPanel headerPanel = new JPanel(new BorderLayout());
		headerPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		headerPanel.setBorder(new EmptyBorder(10, 10, 10, 10));

		enabledBox.setToolTipText("Enable or disable this camera point group");
		enabledBox.setSelected(group.isEnabled());
		enabledBox.addItemListener(e -> {
			group.setEnabled(enabledBox.isSelected());
			plugin.updateConfig();
		});
		headerPanel.add(enabledBox, BorderLayout.WEST);

		JPanel namePanel = new JPanel(new BorderLayout());
		namePanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		namePanel.setBorder(new CompoundBorder(
			new MatteBorder(0, 0, 1, 0, ColorScheme.DARK_GRAY_COLOR),
			new LineBorder(ColorScheme.DARKER_GRAY_COLOR)));

		JPanel nameActions = new JPanel(new BorderLayout(4, 0));
		nameActions.setBackground(ColorScheme.DARKER_GRAY_COLOR);

		configureNameEditActions(nameActions);

		nameInput.setText(group.getName());
		nameInput.setEditable(false);
		nameInput.getTextField().setForeground(Color.WHITE);
		nameInput.getTextField().setFont(FontManager.getRunescapeBoldFont());
		nameInput.getTextField().setBorder(new EmptyBorder(0, 8, 0, 0));
		nameInput.addKeyListener(new KeyAdapter()
		{
			@Override
			public void keyPressed(KeyEvent e)
			{
				if (e.getKeyCode() == KeyEvent.VK_ENTER)
				{
					saveName();
				}
				else if (e.getKeyCode() == KeyEvent.VK_ESCAPE)
				{
					cancelRename();
				}
			}
		});
		nameInput.getTextField().addMouseListener(new MouseAdapter()
		{
			@Override
			public void mouseClicked(MouseEvent e)
			{
				if (!Helper.checkClick(e) || e.getClickCount() < 2 || !renameLabel.isVisible())
				{
					return;
				}
				startRename();
			}
		});

		namePanel.add(nameActions, BorderLayout.EAST);
		namePanel.add(nameInput, BorderLayout.CENTER);
		headerPanel.add(namePanel, BorderLayout.CENTER);

		JPanel actionsPanel = new JPanel(new GridBagLayout());
		actionsPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);

		GridBagConstraints keybindConstraints = new GridBagConstraints();
		keybindConstraints.fill = GridBagConstraints.HORIZONTAL;
		keybindConstraints.weightx = 0.4;
		keybindConstraints.insets = new Insets(0, 2, 0, 2);

		GridBagConstraints itemConstraints = new GridBagConstraints();
		itemConstraints.fill = GridBagConstraints.HORIZONTAL;
		itemConstraints.weightx = 0.2;
		itemConstraints.insets = new Insets(0, 2, 0, 2);

		configureKeybindButton(prevKeybind, group.getPreviousKeybind(), "Load previous camera point in group", keybind -> group.setPreviousKeybind(keybind));
		actionsPanel.add(prevKeybind, keybindConstraints);

		configureKeybindButton(nextKeybind, group.getNextKeybind(), "Load next camera point in group", keybind -> group.setNextKeybind(keybind));
		actionsPanel.add(nextKeybind, keybindConstraints);

		addPointLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		addPointLabel.setToolTipText("Add a camera point to this group");
		addPointLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				addPointLabel.setIcon(addPressedIcon);
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				plugin.addPointToGroup(group);
				plugin.updateConfig();
				reload();
				addPointLabel.setIcon(addHoverIcon);
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				addPointLabel.setIcon(addHoverIcon);
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				addPointLabel.setIcon(addIcon);
			}
		});
		actionsPanel.add(addPointLabel, itemConstraints);

		deleteLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		deleteLabel.setToolTipText("Delete this camera point group");
		deleteLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				deleteLabel.setIcon(deletePressedIcon);
			}

			@Override
			public void mouseReleased(MouseEvent e)
			{
				if (!Helper.checkClick(e))
				{
					return;
				}
				int result = JOptionPane.showConfirmDialog(GroupPanel.this,
					"Are you sure you want to delete this camera point group?",
					"Are you sure?",
					JOptionPane.OK_CANCEL_OPTION);
				if (result == 0)
				{
					plugin.getCameraPointGroupManager().removeGroup(group);
					plugin.updateConfig();
					reloadFunction.run();
				}
				deleteLabel.setIcon(deleteHoverIcon);
			}

			@Override
			public void mouseEntered(MouseEvent e)
			{
				deleteLabel.setIcon(deleteHoverIcon);
			}

			@Override
			public void mouseExited(MouseEvent e)
			{
				deleteLabel.setIcon(deleteIcon);
			}
		});
		actionsPanel.add(deleteLabel, itemConstraints);

		headerPanel.add(actionsPanel, BorderLayout.SOUTH);
		add(headerPanel, BorderLayout.NORTH);

		pointsContainer = new JPanel();
		pointsContainer.setLayout(new BoxLayout(pointsContainer, BoxLayout.Y_AXIS));
		add(pointsContainer, BorderLayout.CENTER);
		reload();
	}

	private void configureNameEditActions(JPanel nameActions)
	{
		saveLabel.setVisible(false);
		saveLabel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		saveLabel.setFont(FontManager.getRunescapeSmallFont());
		saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR);
		saveLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.brighter());
			}

			@Override
			public void mouseReleased(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				saveName();
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.darker());
			}

			@Override
			public void mouseEntered(MouseEvent mouseEvent)
			{
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR.darker());
			}

			@Override
			public void mouseExited(MouseEvent mouseEvent)
			{
				saveLabel.setForeground(ColorScheme.PROGRESS_COMPLETE_COLOR);
			}
		});
		nameActions.add(saveLabel, BorderLayout.EAST);

		cancelLabel.setVisible(false);
		cancelLabel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		cancelLabel.setFont(FontManager.getRunescapeSmallFont());
		cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR);
		cancelLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.brighter());
			}

			@Override
			public void mouseReleased(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				cancelRename();
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.darker());
			}

			@Override
			public void mouseEntered(MouseEvent mouseEvent)
			{
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR.darker());
			}

			@Override
			public void mouseExited(MouseEvent mouseEvent)
			{
				cancelLabel.setForeground(ColorScheme.PROGRESS_ERROR_COLOR);
			}
		});
		nameActions.add(cancelLabel, BorderLayout.WEST);

		renameLabel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
		renameLabel.setFont(FontManager.getRunescapeSmallFont());
		renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker());
		renameLabel.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mousePressed(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR);
			}

			@Override
			public void mouseReleased(MouseEvent mouseEvent)
			{
				if (!Helper.checkClick(mouseEvent))
				{
					return;
				}
				startRename();
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker().darker());
			}

			@Override
			public void mouseEntered(MouseEvent mouseEvent)
			{
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker().darker());
			}

			@Override
			public void mouseExited(MouseEvent mouseEvent)
			{
				renameLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR.darker());
			}
		});
		nameActions.add(renameLabel, BorderLayout.CENTER);
	}

	private interface KeybindSetter
	{
		void set(Keybind keybind);
	}

	private void configureKeybindButton(JButton button, Keybind keybind, String toolTip, KeybindSetter setter)
	{
		button.setPreferredSize(new Dimension(73, 22));
		button.setMinimumSize(new Dimension(73, 0));
		button.setMaximumSize(new Dimension(73, Integer.MAX_VALUE));
		button.setText(keybind.toString());
		button.setToolTipText(toolTip);
		button.setFont(FontManager.getDefaultFont().deriveFont(12f));
		button.addMouseListener(new MouseAdapter()
		{
			@Override
			public void mouseReleased(MouseEvent mouseEvent)
			{
				button.setText(Keybind.NOT_SET.toString());
				setter.set(Keybind.NOT_SET);
				plugin.updateConfig();
			}
		});
		button.addKeyListener(new KeyAdapter()
		{
			@Override
			public void keyPressed(KeyEvent e)
			{
				if (e.getKeyCode() == KeyEvent.VK_ESCAPE)
				{
					requestFocusInWindow();
					return;
				}
				if (Helper.isModifierKey(e.getKeyCode()))
				{
					return;
				}
				Keybind hotkey = new Keybind(e);
				button.setText(hotkey.toString());
				setter.set(hotkey);
				plugin.updateConfig();
				requestFocusInWindow();
			}
		});
	}

	private void startRename()
	{
		nameInput.setEditable(true);
		saveLabel.setVisible(true);
		cancelLabel.setVisible(true);
		renameLabel.setVisible(false);
		nameInput.getTextField().requestFocusInWindow();
		nameInput.getTextField().selectAll();
	}

	private void saveName()
	{
		nameInput.setEditable(false);
		group.setName(nameInput.getText());
		plugin.updateConfig();
		saveLabel.setVisible(false);
		cancelLabel.setVisible(false);
		renameLabel.setVisible(true);
		requestFocusInWindow();
	}

	private void cancelRename()
	{
		nameInput.setEditable(false);
		nameInput.setText(group.getName());
		saveLabel.setVisible(false);
		cancelLabel.setVisible(false);
		renameLabel.setVisible(true);
		requestFocusInWindow();
	}

	public void reload()
	{
		pointsContainer.removeAll();
		for (CameraPoint point : plugin.getCameraPointGroupManager().getAllPointsForGroup(group))
		{
			PointPanel pointPanel = new PointPanel(plugin, group, point, reloadFunction);
			pointsContainer.add(pointPanel);
			pointsContainer.add(Box.createVerticalStrut(5));
		}
		revalidate();
		repaint();
	}
}

File: src/main/java/com/camerapoints/ui/GroupPanel.kt
package com.camerapoints.ui

import com.camerapoints.CameraPointsPlugin
import com.camerapoints.models.CameraPointGroup
import net.runelite.client.config.Keybind
import net.runelite.client.ui.ColorScheme
import net.runelite.client.ui.FontManager
import net.runelite.client.ui.components.FlatTextField
import net.runelite.client.util.ImageUtil
import java.awt.*
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.*
import javax.swing.border.CompoundBorder
import javax.swing.border.EmptyBorder
import javax.swing.border.LineBorder
import javax.swing.border.MatteBorder

class GroupPanel(val plugin: CameraPointsPlugin, private val group: CameraPointGroup, private val reloadFunction: () -> Unit) : JPanel() {

    val ADD_ICON = ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "add_icon.png"))
    val ADD_HOVER_ICON = ImageIcon(ImageUtil.alphaOffset(ADD_ICON.image, -100))
    val ADD_PRESSED_ICON = ImageIcon(ImageUtil.alphaOffset(ADD_ICON.image, -50))

    val DELETE_ICON = ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "delete_icon.png"))
    val DELETE_HOVER_ICON =
        ImageIcon(ImageUtil.loadImageResource(CameraPointsPlugin::class.java, "delete_icon_red.png"))
    val DELETE_PRESSED_ICON = ImageIcon(ImageUtil.alphaOffset(DELETE_HOVER_ICON.image, -50))

    val enabledBox: JCheckBox = JCheckBox()
    val nameInput: FlatTextField = FlatTextField()
    val saveLabel: JLabel = JLabel("Save")
    val cancelLabel: JLabel = JLabel("Cancel")
    val renameLabel: JLabel = JLabel("Rename")
    val addPointLabel: JLabel = JLabel(ADD_ICON)
    val prevKeybind: JButton = JButton()
    val nextKeybind: JButton = JButton()
    val deleteLabel: JLabel = JLabel(DELETE_ICON)
    private val pointsContainer: JPanel

    init {
        layout = BorderLayout()
        border = EmptyBorder(5, 5, 5, 5)
        background = ColorScheme.DARKER_GRAY_COLOR

        // Header Panel
        val headerPanel = JPanel(BorderLayout())
        headerPanel.background = ColorScheme.DARKER_GRAY_COLOR
        headerPanel.border = EmptyBorder(10, 10, 10, 10)

        // Row 1: Checkbox, Name, Plus
        enabledBox.toolTipText = "Enable or disable this camera point group"
        enabledBox.isSelected = group.enabled
        enabledBox.addItemListener {
            group.enabled = enabledBox.isSelected
            plugin.updateConfig()
        }
        headerPanel.add(enabledBox, BorderLayout.WEST)

        val namePanel = JPanel(BorderLayout())
        namePanel.background = ColorScheme.DARKER_GRAY_COLOR
        namePanel.border = CompoundBorder(
            MatteBorder(0, 0, 1, 0, ColorScheme.DARK_GRAY_COLOR),
            LineBorder(ColorScheme.DARKER_GRAY_COLOR)
        )

        val nameActions = JPanel(BorderLayout(4, 0))
        nameActions.background = ColorScheme.DARKER_GRAY_COLOR

        saveLabel.isVisible = false
        saveLabel.background = ColorScheme.DARKER_GRAY_COLOR
        saveLabel.font = FontManager.getRunescapeSmallFont()
        saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR
        saveLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.brighter()
            }
            override fun mouseReleased(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                nameInput.setEditable(false)
                group.name = nameInput.text
                plugin.updateConfig()
                saveLabel.isVisible = false
                cancelLabel.isVisible = false
                renameLabel.isVisible = true
                requestFocusInWindow()
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.darker()
            }
            override fun mouseEntered(mouseEvent: MouseEvent) {
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR.darker()
            }
            override fun mouseExited(mouseEvent: MouseEvent) {
                saveLabel.foreground = ColorScheme.PROGRESS_COMPLETE_COLOR
            }
        })
        nameActions.add(saveLabel, BorderLayout.EAST)

        cancelLabel.isVisible = false
        cancelLabel.background = ColorScheme.DARKER_GRAY_COLOR
        cancelLabel.font = FontManager.getRunescapeSmallFont()
        cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR
        cancelLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.brighter()
            }
            override fun mouseReleased(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                nameInput.setEditable(false)
                nameInput.text = group.name
                saveLabel.isVisible = false
                cancelLabel.isVisible = false
                renameLabel.isVisible = true
                requestFocusInWindow()
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.darker()
            }
            override fun mouseEntered(mouseEvent: MouseEvent) {
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR.darker()
            }
            override fun mouseExited(mouseEvent: MouseEvent) {
                cancelLabel.foreground = ColorScheme.PROGRESS_ERROR_COLOR
            }
        })
        nameActions.add(cancelLabel, BorderLayout.WEST)

        renameLabel.background = ColorScheme.DARKER_GRAY_COLOR
        renameLabel.font = FontManager.getRunescapeSmallFont()
        renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker()
        renameLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR
            }
            override fun mouseReleased(mouseEvent: MouseEvent) {
                if (!Helper.checkClick(mouseEvent))
                    return
                nameInput.setEditable(true)
                saveLabel.isVisible = true
                cancelLabel.isVisible = true
                renameLabel.isVisible = false
                nameInput.textField.requestFocusInWindow()
                nameInput.textField.selectAll()
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker().darker()
            }
            override fun mouseEntered(mouseEvent: MouseEvent) {
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker().darker()
            }
            override fun mouseExited(mouseEvent: MouseEvent) {
                renameLabel.foreground = ColorScheme.LIGHT_GRAY_COLOR.darker()
            }
        })
        nameActions.add(renameLabel, BorderLayout.CENTER)

        namePanel.add(nameActions, BorderLayout.EAST)

        nameInput.text = group.name
        nameInput.setEditable(false)
        nameInput.textField.foreground = Color.WHITE
        nameInput.textField.font = FontManager.getRunescapeBoldFont()
        nameInput.textField.border = EmptyBorder(0, 8, 0, 0)
        nameInput.addKeyListener(object: KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode == KeyEvent.VK_ENTER) {
                    nameInput.setEditable(false)
                    group.name = nameInput.text
                    plugin.updateConfig()
                    saveLabel.isVisible = false
                    cancelLabel.isVisible = false
                    renameLabel.isVisible = true
                    requestFocusInWindow()
                } else if (e.keyCode == KeyEvent.VK_ESCAPE) {
                    nameInput.setEditable(false)
                    nameInput.text = group.name
                    saveLabel.isVisible = false
                    cancelLabel.isVisible = false
                    renameLabel.isVisible = true
                    requestFocusInWindow()
                }
            }
        })
        nameInput.textField.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (!Helper.checkClick(e) || e.clickCount < 2 || !renameLabel.isVisible)
                    return
                nameInput.setEditable(true)
                saveLabel.isVisible = true
                cancelLabel.isVisible = true
                renameLabel.isVisible = false
                nameInput.textField.requestFocusInWindow()
                nameInput.textField.selectAll()
            }
        })
        namePanel.add(nameInput, BorderLayout.CENTER)

        headerPanel.add(namePanel, BorderLayout.CENTER)

        val actionsPanel = JPanel(GridBagLayout())
        actionsPanel.background = ColorScheme.DARKER_GRAY_COLOR

        val keybindConstraints = GridBagConstraints()
        keybindConstraints.fill = GridBagConstraints.HORIZONTAL
        keybindConstraints.weightx = 0.4
        keybindConstraints.insets = Insets(0, 2, 0, 2)

        val itemConstraints = GridBagConstraints()
        itemConstraints.fill = GridBagConstraints.HORIZONTAL
        itemConstraints.weightx = 0.2
        itemConstraints.insets = Insets(0, 2, 0, 2)

        // Row 2: Hotkey buttons and X
        prevKeybind.preferredSize = Dimension(73, 22)
        prevKeybind.minimumSize = Dimension(73, 0)
        prevKeybind.maximumSize = Dimension(73, Int.MAX_VALUE)
        prevKeybind.text = group.previousKeybind.toString()
        prevKeybind.toolTipText = "Load previous camera point in group"
        prevKeybind.font = FontManager.getDefaultFont().deriveFont(12f)
        prevKeybind.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(mouseEvent: MouseEvent) {
                prevKeybind.text = Keybind.NOT_SET.toString()
                group.previousKeybind = Keybind.NOT_SET
                plugin.updateConfig()
            }
        })
        prevKeybind.addKeyListener(object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode == KeyEvent.VK_ESCAPE) {
                    requestFocusInWindow()
                    return
                }
                val hotkey = Keybind(e)
                prevKeybind.text = hotkey.toString()
                group.previousKeybind = hotkey
                plugin.updateConfig()
                requestFocusInWindow()
            }
        })
        actionsPanel.add(prevKeybind, keybindConstraints)

        nextKeybind.preferredSize = Dimension(73, 22)
        nextKeybind.minimumSize = Dimension(73, 0)
        nextKeybind.maximumSize = Dimension(73, Int.MAX_VALUE)
        nextKeybind.text = group.nextKeybind.toString()
        nextKeybind.toolTipText = "Load next camera point in group"
        nextKeybind.font = FontManager.getDefaultFont().deriveFont(12f)
        nextKeybind.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(mouseEvent: MouseEvent) {
                nextKeybind.text = Keybind.NOT_SET.toString()
                group.nextKeybind = Keybind.NOT_SET
                plugin.updateConfig()
            }
        })
        nextKeybind.addKeyListener(object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode == KeyEvent.VK_ESCAPE) {
                    requestFocusInWindow()
                    return
                }
                val hotkey = Keybind(e)
                nextKeybind.text = hotkey.toString()
                group.nextKeybind = hotkey
                plugin.updateConfig()
                requestFocusInWindow()
            }
        })
        actionsPanel.add(nextKeybind, keybindConstraints)

        addPointLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
        addPointLabel.toolTipText = "Add a camera point to this group"
        addPointLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                addPointLabel.icon = ADD_PRESSED_ICON
            }
            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                plugin.cameraPointGroupManager.addPointToGroup(group)
                plugin.updateConfig()
                reload()
                addPointLabel.icon = ADD_HOVER_ICON
            }
            override fun mouseEntered(e: MouseEvent?) {
                addPointLabel.icon = ADD_HOVER_ICON
            }
            override fun mouseExited(e: MouseEvent?) {
                addPointLabel.icon = ADD_ICON
            }
        })
        actionsPanel.add(addPointLabel, itemConstraints)

        deleteLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
        deleteLabel.toolTipText = "Delete this camera point group"
        deleteLabel.addMouseListener(object : MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                deleteLabel.icon = DELETE_PRESSED_ICON
            }

            override fun mouseReleased(e: MouseEvent) {
                if (!Helper.checkClick(e))
                    return
                val result = JOptionPane.showConfirmDialog(this@GroupPanel, "Are you sure you want to delete this camera point group?", "Are you sure?", JOptionPane.OK_CANCEL_OPTION)
                if (result == 0) {
                    plugin.cameraPointGroupManager.removeGroup(group)
                    plugin.updateConfig()
                    reloadFunction()
                }
                deleteLabel.icon = DELETE_HOVER_ICON
            }

            override fun mouseEntered(e: MouseEvent) {
                deleteLabel.icon = DELETE_HOVER_ICON
            }

            override fun mouseExited(e: MouseEvent) {
                deleteLabel.icon = DELETE_ICON
            }
        })
        actionsPanel.add(deleteLabel, itemConstraints)

        headerPanel.add(actionsPanel, BorderLayout.SOUTH)

        add(headerPanel, BorderLayout.NORTH)

        // Points Container
        pointsContainer = JPanel()
        pointsContainer.layout = BoxLayout(pointsContainer, BoxLayout.Y_AXIS)
        add(pointsContainer, BorderLayout.CENTER)


        // Load initial points
        reload()
    }

    fun reload() {
        pointsContainer.removeAll()

        for (point in plugin.cameraPointGroupManager.getAllPointsForGroup(group)) {
            val pointPanel = PointPanel(plugin, group, point, reloadFunction)
            pointsContainer.add(pointPanel)
            pointsContainer.add(Box.createVerticalStrut(5))
        }

        revalidate()
        repaint()
    }
}